src/R/Editor/Impl/Data/WorkspaceVariableProvider.cs

```diff
@@ -84,54 +84,25 @@ internal sealed class WorkspaceVariableProvider : RSessionChangeWatcher, IVariab
-                        if (children != null) {
-                            eval = children.FirstOrDefault((x) => x != null && x.Name == part && !x.IsHidden);
-                            if (eval == null) {
-                                break;
-                            }
-                        }
-                    }
-
-                    if (eval != null) {
-                        var children = eval.GetChildrenAsync().WaitTimeout(_maxWaitTime);   // TODO: discuss wait is fine here. If not, how to fix?
-                        if (children != null) {
-                            return children
-                                    .Where(x => !x.IsHidden)
-                                    .Take(maxCount)
-                                    .Select(m => new VariableInfo(m))
-                                    .ToArray();
-                        }
-                    }
-                } else {
-                    // May be a package object line mtcars$
-                    variableName = TrimTrailingSelector(variableName);
-                    var sessionProvider = EditorShell.Current.ExportProvider.GetExportedValue<IRSessionProvider>();
-                    var session = sessionProvider.GetOrCreate(GuidList.InteractiveWindowRSessionGuid);
-                    var infoList = session.DescribeChildrenAsync("NULL", variableName, REvaluationResultProperties.HasChildrenProperty, 
-                                RValueRepresentations.Str(), _maxResults).WaitTimeout(_maxWaitTime);   // TODO: discuss wait is fine here. If not, how to fix?;
-
-                    if (infoList != null) {
-                        return infoList
-                                    .Where(m => m.Name.StartsWithOrdinal("$"))
-                                    .Take(maxCount)
-                                    .Select(m => new VariableInfo(TrimLeadingSelector(m.Name), string.Empty))
-                                    .ToArray();
-                    }
+                // May be a package object line mtcars$
+                variableName = TrimTrailingSelector(variableName);
+                var sessionProvider = EditorShell.Current.ExportProvider.GetExportedValue<IRSessionProvider>();
+                var session = sessionProvider.GetOrCreate(GuidList.InteractiveWindowRSessionGuid);
+                IReadOnlyList<IREvaluationResultInfo> infoList = null;
+                try {
+                    infoList = session.DescribeChildrenAsync(REnvironments.GlobalEnv, variableName, REvaluationResultProperties.HasChildrenProperty,
+                            RValueRepresentations.Str(), _maxResults).WaitTimeout(_maxWaitTime);
+                } catch(TimeoutException) { }
+
+                if (infoList != null) {
+                    return infoList
+                                .Where(m => m.Name.StartsWithOrdinal("$") || m.Name.StartsWithOrdinal("@"))
```

Comment:
This should probably only show one or the other, depending on whether the user has typed `$` or `@` to get here. Also, take a look at `IRValueInfo.AccessorKind`.